go/tools/gazelle/merger/fix.go

```diff
@@ -325,16 +378,30 @@ func FixLoads(oldFile *bf.File) *bf.File {
 	return &fixedFile
 }
 
-// knownKinds is the set of symbols that Gazelle ever generated loads for,
-// including symbols it no longer uses (e.g., cgo_library), not including
-// other symbols loaded manually (e.g., go_embed_data). This function will
-// only add or remove loads of these symbols.
-var knownKinds = map[string]bool{
-	"cgo_library": true,
-	"go_binary":   true,
-	"go_library":  true,
-	"go_prefix":   true,
-	"go_test":     true,
+// knownFiles is the set of labels for files that Gazelle loads symbols from.
+var knownFiles []string
+
+// knownKinds is a map from symbols to labels of the files they are loaded
+// from. All symbols Gazelle ever generated loads for are present, including
+// symbols it no longer uses (e.g., cgo_library). Manually loaded symbols
+// (e.g., go_embed_data) are not included.
+var knownKinds map[string]string
+
+func init() {
+	knownKinds = make(map[string]string)
+	for _, f := range []struct {
+		file  string
+		kinds []string
+	}{
+		{"go:def.bzl", []string{"cgo_library", "go_binary", "go_library", "go_prefix", "go_test"}},
+		{"proto:def.bzl", []string{"go_grpc_library", "go_proto_library"}},
+	} {
+		filename := fmt.Sprintf("@%s//%s", config.RulesGoRepoName, f.file)
```

Comment:
Is it really worth doing this in code rather than in the declaration?
I am just thinking of when we decide to add rules_cc and rules_proto for skylark versions of the native rules that come from separate repositories...

Reply:
Nope, not worth it. Done.

I don't think this architecture will make sense if we extend Gazelle beyond Go and proto. We'll probably want some kind of plugin architecture so that another tool can implement the analog of `packages` and `rules`. The fixing and merging logic is still pretty language-agnostic, but may need a few hints (what files and symbols to load from).

Let's cross that bridge when we come to it though.